### DISCUSSION TITLE
Something like an importmap but for assets?

### DISCUSSION THREAD

**Discussion by trusktr** [WICG/admin#196] [Ideas]
The ability of an `importmap` to map ES modules to URLs where they come from is novel. This helps tooling evolve such that bundlers can more easily analyze a website (f.e. by finding an importmap in the HTML) in order to bundle it, makes managing dependencies easy for developers, etc.

What about assets? For example, let's say that I publish an example that pulls in a images from various domains. Someone would like to fork this example, and without changing all the locations that use the image, would like to map the images to local versions of the images.

They might be able to do this with something like a `<script type="urlmap">`:

```html
<!-- override some images: -->
<script type="urlmap">
{
  "urlmap": {
    "https://example.com/foo.png": "./other.png"
  }
}
<script>

<!-- Or for example override an HTML URL so that no matter how it is triggered (by an anchor, or by a script) it will be mapped to the new URL -->
<script type="urlmap">
{
  "urlmap": {
    "https://example.com/some-page.html": "./other-page.html"
  }
}
<script>

<!-- A wildcard could be powerful: for example swap all pages for some folder: -->
<script type="urlmap">
{
  "urlmap": {
    "https://example.com/foo/*.html": "./bar/*.html"
  }
}
<script>
```

Would something like this be useful?

Seems like it could be, because

- a developer could write a urlmap in a single place, without touching any other app code, and control how the app is linked.
- tooling could look at a urlmap as a source of truth for where to get assets to then cache them locally, without inspecting all source code.
- For non-analyzable (dynamic script) code, a dev could provide a wildcard that a tool could then act upon.
- etc?

**Comment by yoavweiss** [comment]
I believe @jakearchibald had ideas around [Fetch Maps](https://github.com/WICG/import-maps/issues/211) many years ago. This sounds similar.. Maybe he can chime in with historical artifacts and/or context around what happened there.

**Comment by jakearchibald** [comment]
I don't really have any context other than that issue and https://discourse.wicg.io/t/proposal-fetch-maps/4259/.

I felt like the proposal was kinda ignored because it would delay import maps from landing.
